flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Pattern.java

```diff
@@ -129,6 +141,23 @@ public Time getWindowTime() {
 	}
 
 	/**
+	 * Set a matching behaviour that defines if the same event can be used in multiple matching sequences.
+	 *
+	 * @param matchingBehaviour New matching behaviour
+	 * @return The same pattern operator with the specified matching behaviour
+	 */
+	public Pattern<T, F> matchingBehaviour(MatchingBehaviour matchingBehaviour) {
+
+		Pattern<T, F> pattern = this;
+		while (pattern != null) {
+			pattern.matchingBehaviour = matchingBehaviour;
+			pattern = (Pattern<T, F>) pattern.getPrevious();
+		}
+
+		return this;
```

Comment:
Looking at the necessary implementation to wire the matching behaviour through all `Pattern` instances, I wonder whether the `MatchingBehaviour` should rather be given to the `CEP.pattern` call. There it could be given to the generated `NFA` which is responsible for making sure that the behaviour is respected. The API would then look like:

`CEP.pattern(input, pattern, matchingBehaviour)`

What do you think?